File: framework/wcf/Extensibility/Transactions/TransactionMessagePropertyUdpTransport/VB/UdpTransport/UdpInputChannel.vb
'  Copyright (c) Microsoft Corporation. All rights reserved.

Imports System.ServiceModel
Imports System.ServiceModel.Channels

Namespace Microsoft.ServiceModel.Samples
	Friend Class UdpInputChannel
		Inherits ChannelBase
		Implements IInputChannel
		Private encoder As MessageEncoder
		Private messageQueue As InputQueue(Of Message)

		Friend Sub New(ByVal listener As UdpChannelListener)
			MyBase.New(listener)
			Me.encoder = listener.MessageEncoderFactory.Encoder
			Me.messageQueue = New InputQueue(Of Message)()
		End Sub

		Public ReadOnly Property LocalAddress() As EndpointAddress Implements IInputChannel.LocalAddress
			Get
				Return Nothing
			End Get
		End Property

		'Hands the message off to other components higher up the
		'channel stack that have previously called BeginReceive() 
		'and are waiting for messages to arrive on this channel.
		Friend Sub Dispatch(ByVal message As Message)
			Me.messageQueue.EnqueueAndDispatch(message)
		End Sub

       

		'Closes the channel ungracefully during error conditions.
		Protected Overrides Sub OnAbort()
			Me.messageQueue.Close()
		End Sub

		'Closes the channel gracefully during normal conditions.
		Protected Overrides Sub OnClose(ByVal timeout As TimeSpan)
			Me.messageQueue.Close()
		End Sub

        Protected Overrides Function OnBeginClose(ByVal timeout As TimeSpan, ByVal callback As AsyncCallback,
                                                  ByVal state As Object) As IAsyncResult
            Me.OnClose(timeout)
            Return New CompletedAsyncResult(callback, state)
        End Function

		Protected Overrides Sub OnEndClose(ByVal result As IAsyncResult)
			CompletedAsyncResult.End(result)
		End Sub

		Protected Overrides Sub OnOpen(ByVal timeout As TimeSpan)
		End Sub

        Protected Overrides Function OnBeginOpen(ByVal timeout As TimeSpan, ByVal callback As AsyncCallback,
                                                 ByVal state As Object) As IAsyncResult
            Return New CompletedAsyncResult(callback, state)
        End Function

		Protected Overrides Sub OnEndOpen(ByVal result As IAsyncResult)
			CompletedAsyncResult.End(result)
		End Sub

		Public Function Receive() As Message Implements IInputChannel.Receive
			Return Me.Receive(Me.DefaultReceiveTimeout)
		End Function

		Public Function Receive(ByVal timeout As TimeSpan) As Message Implements IInputChannel.Receive
            Dim message As Message = Nothing
			If Me.TryReceive(timeout, message) Then
				Return message
			Else
				Throw CreateReceiveTimedOutException(Me, timeout)
			End If
		End Function

        Public Function BeginReceive(ByVal callback As AsyncCallback,
                                     ByVal state As Object) As IAsyncResult Implements IInputChannel.BeginReceive
            Return Me.BeginReceive(Me.DefaultReceiveTimeout, callback, state)
        End Function

        Public Function BeginReceive(ByVal timeout As TimeSpan, ByVal callback As AsyncCallback,
                                     ByVal state As Object) As IAsyncResult Implements IInputChannel.BeginReceive
            Return Me.BeginTryReceive(timeout, callback, state)
        End Function

		Public Function EndReceive(ByVal result As IAsyncResult) As Message Implements IInputChannel.EndReceive
			Return Me.messageQueue.EndDequeue(result)
		End Function

        Public Function TryReceive(ByVal timeout As TimeSpan,
                                   <System.Runtime.InteropServices.Out()> ByRef message As Message) As Boolean Implements IInputChannel.TryReceive
            UdpChannelHelpers.ValidateTimeout(timeout)
            Return Me.messageQueue.Dequeue(timeout, message)
        End Function

        Public Function BeginTryReceive(ByVal timeout As TimeSpan,
                                        ByVal callback As AsyncCallback,
                                        ByVal state As Object) As IAsyncResult Implements IInputChannel.BeginTryReceive
            UdpChannelHelpers.ValidateTimeout(timeout)
            Return Me.messageQueue.BeginDequeue(timeout, callback, state)
        End Function

        Public Function EndTryReceive(ByVal result As IAsyncResult,
                                      <System.Runtime.InteropServices.Out()> ByRef message As Message) As Boolean Implements IInputChannel.EndTryReceive
            Return Me.messageQueue.EndDequeue(result, message)
        End Function

		Public Function WaitForMessage(ByVal timeout As TimeSpan) As Boolean Implements IInputChannel.WaitForMessage
			UdpChannelHelpers.ValidateTimeout(timeout)
			Return Me.messageQueue.WaitForItem(timeout)
		End Function

        Public Function BeginWaitForMessage(ByVal timeout As TimeSpan,
                                            ByVal callback As AsyncCallback,
                                            ByVal state As Object) As IAsyncResult Implements IInputChannel.BeginWaitForMessage
            UdpChannelHelpers.ValidateTimeout(timeout)
            Return Me.messageQueue.BeginWaitForItem(timeout, callback, state)
        End Function

		Public Function EndWaitForMessage(ByVal result As IAsyncResult) As Boolean Implements IInputChannel.EndWaitForMessage
			Return Me.messageQueue.EndWaitForItem(result)
		End Function

        Private Shared Function CreateReceiveTimedOutException(ByVal channel As IInputChannel,
                                                               ByVal timeout As TimeSpan) As TimeoutException
            If channel.LocalAddress IsNot Nothing Then
                Return New TimeoutException(String.Format("Receive on local address {0} timed out after {1}. The time allotted to this operation may have been a portion of a longer timeout.",
                                                          channel.LocalAddress.Uri.AbsoluteUri,
                                                          timeout))
            Else
                Return New TimeoutException(String.Format("Receive timed out after {0}. The time allotted to this operation may have been a portion of a longer timeout.",
                                                          timeout))
            End If
        End Function
	End Class
End Namespace

File: framework/wcf/Extensibility/Transport/UdpActivation/CS/UDPActivation/Channels/UdpInputChannel.cs

//  Copyright (c) Microsoft Corporation.  All Rights Reserved.


using System;
using System.ServiceModel;
using System.ServiceModel.Channels;

namespace Microsoft.ServiceModel.Samples
{
    class UdpInputChannel : ChannelBase, IInputChannel
    {
        MessageEncoder encoder;
        InputQueue<Message> messageQueue;

        internal UdpInputChannel(UdpChannelListener listener)
            : base(listener)
        {
            this.encoder = listener.MessageEncoderFactory.Encoder;
            this.messageQueue = new InputQueue<Message>();
        }

        public EndpointAddress LocalAddress
        {
            get
            {
                return null;
            }
        }

        //Hands the message off to other components higher up the
        //channel stack that have previously called BeginReceive() 
        //and are waiting for messages to arrive on this channel.
        internal void Dispatch(Message message)
        {
            this.messageQueue.EnqueueAndDispatch(message);
        }

        public override T GetProperty<T>()
        {
            if (typeof(T) == typeof(IInputChannel))
            {
                return (T)(object)this;
            }

            T messageEncoderProperty = this.encoder.GetProperty<T>();
            if (messageEncoderProperty != null)
            {
                return messageEncoderProperty;
            }

            return base.GetProperty<T>();
        }

        //Closes the channel ungracefully during error conditions.
        protected override void OnAbort()
        {
            this.messageQueue.Close();
        }

        //Closes the channel gracefully during normal conditions.
        protected override void OnClose(TimeSpan timeout)
        {
            this.messageQueue.Close();
        }

        protected override IAsyncResult OnBeginClose(TimeSpan timeout, AsyncCallback callback, object state)
        {
            this.OnClose(timeout);
            return new CompletedAsyncResult(callback, state);
        }

        protected override void OnEndClose(IAsyncResult result)
        {
            CompletedAsyncResult.End(result);
        }

        protected override void OnOpen(TimeSpan timeout)
        {
        }

        protected override IAsyncResult OnBeginOpen(TimeSpan timeout, AsyncCallback callback, object state)
        {
            return new CompletedAsyncResult(callback, state);
        }

        protected override void OnEndOpen(IAsyncResult result)
        {
            CompletedAsyncResult.End(result);
        }

        public Message Receive()
        {
            return this.Receive(this.DefaultReceiveTimeout);
        }

        public Message Receive(TimeSpan timeout)
        {
            Message message;
            if (this.TryReceive(timeout, out message))
            {
                return message;
            }
            else
            {
                throw CreateReceiveTimedOutException(this, timeout);
            }
        }

        public IAsyncResult BeginReceive(AsyncCallback callback, object state)
        {
            return this.BeginReceive(this.DefaultReceiveTimeout, callback, state);
        }

        public IAsyncResult BeginReceive(TimeSpan timeout, AsyncCallback callback, object state)
        {
            return this.BeginTryReceive(timeout, callback, state);
        }

        public Message EndReceive(IAsyncResult result)
        {
            return this.messageQueue.EndDequeue(result);
        }

        public bool TryReceive(TimeSpan timeout, out Message message)
        {
            UdpChannelHelpers.ValidateTimeout(timeout);
            return this.messageQueue.Dequeue(timeout, out message);
        }

        public IAsyncResult BeginTryReceive(TimeSpan timeout, AsyncCallback callback, object state)
        {
            UdpChannelHelpers.ValidateTimeout(timeout);
            return this.messageQueue.BeginDequeue(timeout, callback, state);
        }

        public bool EndTryReceive(IAsyncResult result, out Message message)
        {
            return this.messageQueue.EndDequeue(result, out message);
        }

        public bool WaitForMessage(TimeSpan timeout)
        {
            UdpChannelHelpers.ValidateTimeout(timeout);
            return this.messageQueue.WaitForItem(timeout);
        }

        public IAsyncResult BeginWaitForMessage(TimeSpan timeout, AsyncCallback callback, object state)
        {
            UdpChannelHelpers.ValidateTimeout(timeout);
            return this.messageQueue.BeginWaitForItem(timeout, callback, state);
        }

        public bool EndWaitForMessage(IAsyncResult result)
        {
            return this.messageQueue.EndWaitForItem(result);
        }

        static TimeoutException CreateReceiveTimedOutException(IInputChannel channel, TimeSpan timeout)
        {
            if (channel.LocalAddress != null)
            {
                return new TimeoutException(
                    string.Format("Receive on local address {0} timed out after {1}. The time allotted to this operation may have been a portion of a longer timeout.",
                    channel.LocalAddress.Uri.AbsoluteUri, timeout));
            }
            else
            {
                return new TimeoutException(
                    string.Format("Receive timed out after {0}. The time allotted to this operation may have been a portion of a longer timeout.",
                    timeout));
            }
        }
    }
}

